### ISSUE TITLE
traverse: depth-first walk does not visit everything under one node before moving to another

### ISSUE THREAD

**Issue by bboreham** [gonum/gonum#1595]
### What are you trying to do?

Get a depth-first ordering of nodes

### What did you do?

Apologies if this is me misunderstanding the API.
To illustrate, I want to use this simple graph:

```
      1
    /  \
  2     4
 /
3
```

```
package main

import (
	"fmt"

	"gonum.org/v1/gonum/graph"
	"gonum.org/v1/gonum/graph/simple"
	"gonum.org/v1/gonum/graph/traverse"
)

func main() {
	g := simple.NewDirectedGraph()
	g.AddNode(simple.Node(1))
	g.AddNode(simple.Node(2))
	g.AddNode(simple.Node(3))
	g.AddNode(simple.Node(4))
	g.SetEdge(g.NewEdge(simple.Node(1), simple.Node(2)))
	g.SetEdge(g.NewEdge(simple.Node(2), simple.Node(3)))
	g.SetEdge(g.NewEdge(simple.Node(1), simple.Node(4)))
	t := traverse.DepthFirst{
		Visit: func(node graph.Node) {
			fmt.Print(node.(simple.Node))
		},
	}
	t.Walk(g, simple.Node(1), nil)
	fmt.Println()
}
```

(This times out during compilation at https://play.golang.org/p/LBmE0U_AxG_-)

### What did you expect to happen?

Output 1234 or 1423

### What actually happened?

1243

### What version of Go and Gonum are you using?
```
go version go1.16 linux/amd64

gonum.org/v1/gonum v0.9.0

gonum.org/v1/gonum v0.0.0-20180816165407-929014505bf4/go.mod h1:Y+Yx5eoAFn32cQvJDxZx5Dpnq+c3wtXuadVZAcxbbBo=
gonum.org/v1/gonum v0.8.2/go.mod h1:oe/vMfY3deqTw+1EZJhuvEW2iwGF1bW9wwu7XCu0+v0=
gonum.org/v1/gonum v0.9.0 h1:KSrriwTGNTqNqyR6ZWWvwtInD/kl89sTrKsDfmbk2HU=
gonum.org/v1/gonum v0.9.0/go.mod h1:TZumC3NeyVQskjXqmyWt4S3bINhy7B4eYwW69EbyX+0=```
```

### Does this issue reproduce with the current master?

Yes

**Comment by kortschak** [comment]
Thanks. That's an embarrassingly long-lived bug. Please take a look at #1596.

**Comment by bboreham** [comment]
Thanks for quick response!

For my personal taste there are a lot of unrelated changes in that PR so it is hard to check completely; however it does seem to make the key change of moving `Visit()` from the neighbour loop to the stack-pop, so LGTM.

**Comment by kortschak** [comment]
The implementation was a complete rewrite without reference to the original due to the small size of the method, and the test additions were necessary to ensure that traversal order is being tested (the previous tests did not test order but only node coverage). I think the only unrelated change was fixing 2 typos (related to two typos in the tests being fixed).

This will be in v0.9.1 which hopefully will be released in the not to distant future.

**Comment by kortschak** [comment]
@bboreham This fix is now released in v0.9.1.
